### ISSUE TITLE
Ad-Hoc Rate Limiting for the number of concurrently running jobs

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#990]
Originally reported of Google Code with ID 934.

```
It is often very useful to have a lightweight way for one-off rate-limiting. One example would be evacuating a node but limiting the number of simultaneous instance moves to no overload the replication network.

It'd be convenient to mark jobs so that at most _n_ jobs with the same mark would be running in parallel.

```
Originally added on 2014-08-29 11:41:16 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed in:

Date:   Fri Aug 1 17:27:04 2014 +0200
commit 0fd81c8ed7fd6542510fb0f195bb88e599557eea
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:16 2014 +0200

    Tests: Scheduler: Add tests for ad-hoc rate limiting
    
    The main property of rate-limited jobs is tested (jobs must not be
    scheduled over the limit), as well as reason string parsing and
    properties of helper functions.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 8cc1ddce80cd0d0f7acbd7b81d309ec7b9099d1a
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:15 2014 +0200

    Man: Document ad-hoc rate limiting with --reason option
    
    As defined by doc/design-optables.rst.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 6cd1f9ef355aa6966ee5e23adcbc67ff5c0bfdfa
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:14 2014 +0200

    Scheduler: Implement ad-hoc reason trail rate limiting
    
    This implements the "Ad-Hoc Rate Limiting" part of the design
    specified in doc/design-optables.rst.
    
    The current implementation does not try to cache how many slots are
    free per bucket since the number of running jobs is typically small
    (< 100).
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 9805e173ae2578cdf10b6fd7883ce07cc045fc5f
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:13 2014 +0200

    Scheduler: Add nullJobWithStat
    
    This is mainly for tests.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 7a67fa57827ea7f3fbab1dbc2a507f28866da337
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:12 2014 +0200

    Derive Ord for various types
    
    This is so that we can put `QueuedJob`s in a `Set`.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 1220ba1dfa65c5f6eb72aa76e768c56a72c2ab59
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:10 2014 +0200

    Utils: Add readMaybe
    
    Backported from Text.Read.readMaybe (it is in base-4.6 but we need
    to support older versions).
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 3a0be96a03afb03b760180ef87e23a5cf08a6b70
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:09 2014 +0200

    Scheduler: Add Eq/Show instances
    
    This is so that `Queue` and `JobWithStat` can be used in assertions.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 5c3ddba56df3d1ebb31a4f2e5691de9d89108ea6
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:08 2014 +0200

    Tests: Extract Test.Ganeti.JQueue.Objects module
    
    We want to reuse functions for it for a future 'Test.Ganeti.JQScheduler'.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 038538c350b5eef97d40768960f0a1412fb4be12
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:07 2014 +0200

    Scheduler: Move types into own module
    
    So that we can access them from modules that will be imported into JQScheduler
    (such as job queue filters).
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

commit 73abebda6c146b9530aa6087643dfdebba1a9e6d
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:06 2014 +0200

    Tests: Add implement shrink for `Arbitrary Map`
    
    The implementation was taken from `quickcheck-instances`.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 79c397c2bac98da12b3ffae2ae48a93ee0bf2a69
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:05 2014 +0200

    Tests: Move Arbitrary Map instance to TestCommon
    
    We should switch to the quickcheck-instances package for that, but for now
    TestCommon is certainly a better place for it than OpCodes.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit b7f61104e449c51c7e34cbfe42349417ac816973
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:04 2014 +0200

    Man: Document --reason option
    
    The man page entry for it was forgotten when the feature was added.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 21ad488b91e9e6af3df2eec5e29d24b6645e92ce
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:03 2014 +0200

    Types: Manual Ord NonEmpty instance -> deriving instance
    
    The manually written instance was identical to the derived one
    (for this newtype there's only one instance).
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 9511cd18d8a4bae16525a7577ad836cd168b858a
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:02 2014 +0200

    Scheduler: Haddock selectJobsToRun parameter
    
    The "maximum number of jobs to run" parameter was not documented.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit a16e0600edbf397c2606a2dff7e2eb968b337ede
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:01 2014 +0200

    TestCommon: Update historical module description
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 689ffa0694ddf362d59e1d4965c6807ac0e3549b
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Aug 1 17:27:00 2014 +0200

    Makefile.am: Fix file order
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Petr Pudlak <pudlak@google.com>

```
Originally added on 2014-08-29 11:42:41 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
See also the design doc: http://docs.ganeti.org/ganeti/master/html/design-optables.html#additional-ad-hoc-rate-limiting
```
Originally added on 2014-08-29 11:44:59 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
